### PR TITLE
HSGP improvements

### LINKED ISSUE #7240: BUG: HSGP making wrong out-of-sample predictions when X domain changes
### Describe the issue:

While working on the HSGP tutorial with @bwengals, we discovered out-of-sample predictions are wrong when the range of the X domain changes, because `L`, the parameter on the boundary condition is computed again the new values of X... but shouldn't.
I'll fix that in a PR coming very soon. I'll also make sure the mean-centering of data, necessary to compute `L`, is done under the hood, instead of requiring the user to do it -- which is the current API but is a sneaky source of error

### Reproduceable code example:

```python
x = np.linspace(-5, 5, 100)

with pm.Model() as model:
    # Set mutable data
    X = pm.MutableData("X", x[:, None])

    m = [5]
    c = 5.0

    X_mu = np.mean(x, axis=0)
    Xs = X - X_mu
    
    ### WITH EVAL, NO BUG.  WITHOUT EVAL, BUG
    L = pm.gp.hsgp_approx.set_boundary(Xs, c).eval()
    
    eigvals = pm.gp.hsgp_approx.calc_eigenvalues(L, m, tl=pt)
    phi = pm.gp.hsgp_approx.calc_eigenvectors(Xs, L, eigvals, m, tl=pt)

phi1 = phi.eval()

x_new = np.linspace(-5, 10, 200)

with model:
    pm.set_data({"X": x_new[:, None]})
    
plt.plot(x, phi1, 'k');
plt.plot(x_new, phi.eval(), 'b', alpha=0.5);
```

### Error message:

_No response_

### PyMC version information:

5.12.0

### Context for the issue:

_No response_

### PR DESCRIPTION
**Pull request by AlexAndorra**
Closes #7240 

Working on this in concert with @bwengals for [this example](https://github.com/pymc-devs/pymc-examples/pull/647)

- [x] Fix the [bug in the computation](https://github.com/pymc-devs/pymc/issues/7240) of `L`
- [x] Do mean-centering of `X` values under the hood, instead of requiring it from users
- [x] Enable setting `prior_linearized` with `m` and `c` instead of only `m` and `L`
- [x] `parametrization` is not documented
- [x] Add option to pass in nicely named `coords` to HSGP for the `beta` coefficients
- [x] Make `._m_star` more accessible / user friendly (make a property with no leading underscore so existing code doesn't break), document it, and fix example in `prior_linearized` docstring.
- [x] Get rid of numpy option, `tl=np`.  It's subtly wrong when calculating the eigen stuff and its not used anywhere.
- [x] Add the `approx_params` function to `hsgp_approx` file

## Type of change
- New feature / enhancement
- Bug fix

----
📚 Documentation preview 📚: https://pymc--7335.org.readthedocs.build/en/7335/

### CODE REVIEW

**Comment by AlexAndorra** [comment]
The failing test on Windows isn't related to these changes 🤷‍♂️ 
All the tests are passing

File: pymc/gp/hsgp_approx.py
Comment (juanitorduz) [inline pymc/gp/hsgp_approx.py]: What would happen if `Xs` is always negative? say `[-1, -0.8, -0.6]`? If `c` is positive then `L` will be negative. Shall we keep the `abs`?

File: pymc/gp/hsgp_approx.py
Comment (juanitorduz) [inline pymc/gp/hsgp_approx.py:154]: Shall we raise an error with a friendly error message if `cov_function` does not match the expected ones?

File: pymc/gp/hsgp_approx.py
Comment (juanitorduz) [inline pymc/gp/hsgp_approx.py]: Just a suggestion: what about returning a [namedtuple](https://docs.python.org/3/library/collections.html#collections.namedtuple) so that we do not depend on the special order of the parameters? something like

```python
HSGPParam = namedtuple('HSGPParam ', ['m', 'c', 'S'])
```

File: pymc/gp/hsgp_approx.py
Comment (juanitorduz) [inline pymc/gp/hsgp_approx.py]: Do you want to remove `### CHECH WITH c INSTEAD!!` ?

Comment (juanitorduz) [reply pymc/gp/hsgp_approx.py]: (same below)

File: tests/gp/test_hsgp_approx.py
Comment (juanitorduz) [inline tests/gp/test_hsgp_approx.py]: Can we parametrize the test to make sure is cover the cases when `Xs` is always negative and when it contains negative values and positive ones?

**Review by juanitorduz — COMMENTED** [review]
This looks great! Just some minor suggestions :) 

Maybe I missed but I missed, but is there a test where we verify the mean sustraction works as expected in the out-of-sample case?

File: pymc/gp/hsgp_approx.py
Comment (AlexAndorra) [reply pymc/gp/hsgp_approx.py]: Oooooh, you're right! Adding it back

File: pymc/gp/hsgp_approx.py
Comment (AlexAndorra) [reply pymc/gp/hsgp_approx.py:154]: Good point. Done

File: pymc/gp/hsgp_approx.py
Comment (AlexAndorra) [reply pymc/gp/hsgp_approx.py]: Good idea. Done

File: pymc/gp/hsgp_approx.py
Comment (AlexAndorra) [reply pymc/gp/hsgp_approx.py]: Good catch

File: tests/gp/test_hsgp_approx.py
Comment (AlexAndorra) [reply tests/gp/test_hsgp_approx.py]: Done

**Comment by AlexAndorra** [comment]
Thanks @juanitorduz ! I just pushed all the changes.

> is there a test where we verify the mean sustraction works as expected in the out-of-sample case?

Good point. Added it

File: pymc/gp/hsgp_approx.py
Comment (juanitorduz) [inline pymc/gp/hsgp_approx.py:301]: is this double `=` intentional?

```python
  self._m_star = self.n_basis_vectors = int(np.prod(self._m))
```

**Review by juanitorduz — APPROVED** [review]
Looks great! Just left a comment just to make sure its intentional and not a typo ;)

File: pymc/gp/hsgp_approx.py
Comment (AlexAndorra) [reply pymc/gp/hsgp_approx.py:301]: Yes, it's a double assignment (if I remember the technical term correctly)

Comment (juanitorduz) [reply pymc/gp/hsgp_approx.py:301]: ok! just double checking ;)

File: pymc/gp/hsgp_approx.py
Comment (bwengals) [reply pymc/gp/hsgp_approx.py]: Personally think this is over-engineering, because introducing a new `HSGPParam` type serves no purpose other than to compensate for a perceived defect with the Python language.  I think it was clearer before.

File: pymc/gp/hsgp_approx.py
Comment (bwengals) [inline pymc/gp/hsgp_approx.py]: Hate to delay, and OK if it waits till a later PR, but it'd be better to use the midpoint of X here instead of the mean.  So, 
```python
self._X_center = (pt.max(Xs, axis=0) - pt.min(Xs, axis=0)) / 2
```

File: pymc/gp/hsgp_approx.py
Comment (fonnesbeck) [inline pymc/gp/hsgp_approx.py:476]: Need to undo this spelling error, and the one below.

Comment (fonnesbeck) [inline pymc/gp/hsgp_approx.py:478]: FWIW, either spelling is correct.

File: pymc/gp/hsgp_approx.py
Comment (AlexAndorra) [reply pymc/gp/hsgp_approx.py]: On the contrary, it's great to have your input @bwengals !
Why is that though? For `S` I see why, but for the eigen vectors, all the examples before were using the mean, not the center if I'm not mistaken 🤔

File: pymc/gp/hsgp_approx.py
Comment (AlexAndorra) [reply pymc/gp/hsgp_approx.py:478]: Is it?? Damn, all these changes for nothing then 😅

File: pymc/gp/hsgp_approx.py
Comment (bwengals) [reply pymc/gp/hsgp_approx.py]: yah they were, was a mistake I think.  I think for the same reason as S, so those line up.  Then the accuracy of the approximation doesn't depend on the density of the x locations.

File: pymc/gp/hsgp_approx.py
Comment (juanitorduz) [reply pymc/gp/hsgp_approx.py]: No strong opinion here. I just mess up indices so often that I prefer going safe with names instead 🙈

File: pymc/gp/hsgp_approx.py
Comment (juanitorduz) [reply pymc/gp/hsgp_approx.py]: Wow ... this is an interesting point! Imagine X consists of many points around zero and a single point at 100. If we want the box dimension `L` to be as small as possible (that is the whole aim right?) then we should not take the mean but the mid-point indeed 🤔

Comment (AlexAndorra) [reply pymc/gp/hsgp_approx.py]: Mind blown @bwengals 🤯 But that makes total sense now! Even better that we're doing the transformation under the hood now, otherwise we'd have to change a lot of examples...
Just pushed the changes. Feel free to merge if it looks good now

Comment (bwengals) [reply pymc/gp/hsgp_approx.py]: > Even better that we're doing the transformation under the hood now, otherwise we'd have to change a lot of examples... 

Definitely, awesome that you got that working now

**Review by bwengals — APPROVED** [review]

**Comment by bwengals** [comment]
Looks like there's one failing test mentioning `_X_mean`,

```python
         assert np.allclose(
            gp._X_mean, original_mean
>       ), "gp._X_mean should not change after updating data for out-of-sample predictions."
E       AttributeError: 'HSGP' object has no attribute '_X_mean'
```

File: pymc/gp/hsgp_approx.py
Comment (bwengals) [reply pymc/gp/hsgp_approx.py]: yah but I would just expect people to unpack like, `m, c, S = ...`.  

Which makes me think it might make sense (in this or a later PR) to slightly refactor this so that there's a different function,
```python
S = calc_X_radius(X) # or calc_halfwidth or something
```
that could be used later where `_X_mean` was replaced.  Then this function only needs to return `m` and `c` and not also `S`, since that's handled now.

File: pymc/gp/hsgp_approx.py
Comment (AlexAndorra) [inline pymc/gp/hsgp_approx.py]: Actually, shouldn't this also be using the midpoint @bwengals  (i.e `self._X_center`):
```python
S = (pt.max(Xs, axis=0) - pt.min(Xs, axis=0)) / 2
```
The current formula seems to compute the _largest magnitude_ of X, _not_ its half range, does it?

File: pymc/gp/hsgp_approx.py
Comment (AlexAndorra) [inline pymc/gp/hsgp_approx.py:146]: After talking to @bwengals , I changed the computation here, to make sure it matches the definition of `S` in `set_boundary`.
The previous version was only doing;
```python
S = (x_range[1] - x_range[0]) / 2
```
If that looks good, I think we're good to merge 🥳

Comment (bwengals) [reply pymc/gp/hsgp_approx.py:146]: yah good change.  Hit the merge button!

**Review by bwengals — APPROVED** [review]

File: pymc/gp/hsgp_approx.py
Comment (bwengals) [inline pymc/gp/hsgp_approx.py]: just noticed that this bit of the docstring is out of date.  looks like the function takes x instead of x_range now

Comment (bwengals) [reply pymc/gp/hsgp_approx.py]: Since it's x now, it'd be good to remind people in the docstring that this `x` should be the x values you'll want to predict over, not just the training x.

Comment (AlexAndorra) [reply pymc/gp/hsgp_approx.py]: Good catch @bwengals ! Just pushed the change. Will merge once it's all green

**Pull request merged**
